clients/src/main/java/org/apache/kafka/common/requests/FindCoordinatorRequest.java

```
@@ -0,0 +1,145 @@
+            super(ApiKeys.FIND_COORDINATOR);
+            this.coordinatorType = coordinatorType;
+            this.coordinatorKey = coordinatorKey;
+            this.minVersion = coordinatorType == CoordinatorType.TRANSACTION ? (short) 1 : (short) 0;
+        }
+
+        @Override
+        public FindCoordinatorRequest build(short version) {
+            if (version < minVersion)
+                throw new UnsupportedVersionException("Cannot create a v" + version + " FindCoordinator request " +
+                        "because we require features supported only in " + minVersion + " or later.");
+            return new FindCoordinatorRequest(coordinatorType, coordinatorKey, version);
+        }
+
+        @Override
+        public String toString() {
+            StringBuilder bld = new StringBuilder();
+            bld.append("(type=FindCoordinatorRequest, coordinatorKey=");
+            bld.append(coordinatorKey);
+            bld.append(", coordinatorType=");
+            bld.append(coordinatorType);
+            bld.append(")");
+            return bld.toString();
+        }
+    }
+
+    private final String coordinatorKey;
+    private final CoordinatorType coordinatorType;
+
+    private FindCoordinatorRequest(CoordinatorType coordinatorType, String coordinatorKey, short version) {
+        super(version);
+        this.coordinatorType = coordinatorType;
+        this.coordinatorKey = coordinatorKey;
+    }
+
+    public FindCoordinatorRequest(Struct struct, short version) {
+        super(version);
+
+        if (version == 0) {
+            this.coordinatorType = CoordinatorType.GROUP;
+            this.coordinatorKey = struct.getString(GROUP_ID_KEY_NAME);
+        } else {
+            this.coordinatorType = CoordinatorType.forId(struct.getByte(COORDINATOR_TYPE_KEY_NAME));
+            this.coordinatorKey = struct.getString(COORDINATOR_KEY_KEY_NAME);
+        }
+    }
+
+    @Override
+    public AbstractResponse getErrorResponse(Throwable e) {
```

Discussed offline. Since we aren't using exceptions directly to pass error codes in the response, the error is unexpected (i.e. a bug), so maybe we can pass it back directly to the client and let it be raised to the user.